{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import xlsxwriter\n",
    "import math\n",
    "import Bio\n",
    "import sys\n",
    "import os\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import *\n",
    "from Bio.Alphabet import IUPAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skip the last oligo for RDN5, as it is shorter tha or equal to 6 bp\n"
     ]
    }
   ],
   "source": [
    "\"\"\"DataFrame Initialization\n",
    "One for the original sequence reference\n",
    "and the other for oligos\n",
    "\n",
    "Oligo generation and original sequence reference recording\n",
    "Iterate through the number of sequences \n",
    "to update the original sequence reference dataframe with\n",
    "the sequence ID, the original sequence, \n",
    "its reverse and complementary sequence, and its sequence length \n",
    "\n",
    "For each sequence, iterate per 60 bp \n",
    "to update the oligo dataframe with\n",
    "the target name that is the sequence ID, \n",
    "the probe name that is the sequence ID \n",
    "plus the oligo number for that sequence ID,\n",
    "and the 60 bp or shorter oligo sequence itself \n",
    "in the 5' to 3' direction\n",
    "\n",
    "Note that oligos less than 4 bp are elimiated\n",
    "\n",
    "Export the dataframes to an Excel file\n",
    "One dataframe per sheet with adjustments such as\n",
    "the column widths, alignments, and text fonts.\n",
    "\"\"\"\n",
    "\n",
    "def SDRNA_Oligos(inputfolder, fastafile, strainname, details):\n",
    "    pathname = Path(inputfolder + \"/\")\n",
    "    fastafile = fastafile + \".fasta\"\n",
    "    inputfile = pathname / fastafile\n",
    "    \n",
    "    d = {\"SeqID\": \"\", \"ConSeq\": \"\", \"RevComp\": \"\", \"SeqLen\": []}\n",
    "    df = pd.DataFrame(data=d)\n",
    "\n",
    "    opsheet = {\"Target\": [],\n",
    "               \"Probe Name\": [],\n",
    "               \"Sequence (5'-3')\": []}\n",
    "    opf = pd.DataFrame(data=opsheet)\n",
    "    opname = (\"SDRNA oligos in \" + str(strainname) \n",
    "             + \" \" + str(details))\n",
    "    \n",
    "    opfasta = opname + \".fasta\"\n",
    "    opfastapath = pathname / opfasta\n",
    "    f = open(opfastapath, \"w\")\n",
    "\n",
    "    col1width = 0\n",
    "    col2width = 0 # initialize the column width for the probe name\n",
    "        \n",
    "    for seq_record in SeqIO.parse(inputfile, \"fasta\"):\n",
    "        revcomp = str(seq_record.seq.reverse_complement()[::])\n",
    "        seqlen = len(seq_record)\n",
    "        \n",
    "        targetname = seq_record.id.split(\"-\")[0]\n",
    "        # targetname = seq_record.id.split(\"|\")[4]\n",
    "                           \n",
    "        df = df.append({\"SeqID\": targetname,\n",
    "                        \"ConSeq\": str(seq_record.seq[::]),\n",
    "                        \"RevComp\": revcomp,\n",
    "                        \"SeqLen\": seqlen}, ignore_index=True)\n",
    "            \n",
    "        oligtot = math.ceil(seqlen/60)\n",
    "        \n",
    "        for olignumb in range(oligtot):\n",
    "                                  \n",
    "            seqstart = olignumb*60\n",
    "            \n",
    "            if olignumb != oligtot-1:\n",
    "                seqend = seqstart+60\n",
    "            elif olignumb == oligtot-1 and seqlen-seqstart < 7:\n",
    "                print(\"Skip the last oligo for \" + str(targetname) + \n",
    "                      \", as it is shorter than or equal to 6 bp\")\n",
    "                continue\n",
    "            else:\n",
    "                seqstart = seqlen-60\n",
    "                seqend = seqlen\n",
    "            \n",
    "            probename = str(targetname) + \"-SDRNA-\" + str(olignumb+1)\n",
    "            probeseq = revcomp[seqstart:seqend]\n",
    "            \n",
    "            opf = opf.append({\"Target\": targetname,\n",
    "                              \"Probe Name\": probename,\n",
    "                              \"Sequence (5'-3')\": probeseq}, ignore_index=True)\n",
    "\n",
    "            f.write(\">%s\\n%s\\n\" % (probename,probeseq))\n",
    "         \n",
    "        if len(targetname) > col1width:\n",
    "            col1width = len(targetname)\n",
    "        \n",
    "        if len(probename) > col2width:\n",
    "            col2width = len(probename)\n",
    "            \n",
    "    f.close()\n",
    "    \n",
    "    fastaoligname = \"SDRNA oligos in \" + str(strainname) + \" \" + str(details)\n",
    "    \n",
    "    df[\"SeqLen\"] = df[\"SeqLen\"].astype(\"int\")\n",
    "    \n",
    "    opExcel = opname + \".xlsx\"\n",
    "    opExcelpath = pathname / opExcel\n",
    "    writer = pd.ExcelWriter(opExcelpath, engine='xlsxwriter')\n",
    "    opf.to_excel(writer, sheet_name=\"Oligos\", \n",
    "                 header=True, index=False, index_label=None)\n",
    "    df.to_excel(writer, sheet_name=\"Sequences\", \n",
    "                header=True, index=False, index_label=None)\n",
    "    workbook  = writer.book\n",
    "    leftformat = workbook.add_format({\"text_wrap\": True,\n",
    "                                      \"valign\": \"vcenter\",\n",
    "                                      \"font_name\": \"Courier New\"})\n",
    "    centerformat = workbook.add_format({\"text_wrap\": True,\n",
    "                                        \"align\": \"center\",\n",
    "                                        \"valign\": \"vcenter\",\n",
    "                                        \"font_name\": \"Courier New\"})\n",
    "    \n",
    "    worksheet1 = writer.sheets[\"Oligos\"]\n",
    "    worksheet1.set_column(\"A:A\", col1width*2, leftformat)\n",
    "    worksheet1.set_column(\"B:B\", col2width*2, leftformat)\n",
    "    worksheet1.set_column(\"C:C\", 80, leftformat)\n",
    "    \n",
    "    worksheet2 = writer.sheets[\"Sequences\"]\n",
    "    worksheet2.set_column(\"A:A\", col1width*2, centerformat)\n",
    "    worksheet2.set_column(\"D:D\", None, centerformat)\n",
    "    worksheet2.set_column(\"B:C\", 60, leftformat)\n",
    "\n",
    "    writer.save()\n",
    "    writer.close()\n",
    "    \n",
    "    return fastaoligname\n",
    "    \n",
    "fastaoligname = SDRNA_Oligos(\"NCBI or Manual Compiling Testing/SC Testing\", \"SC Compiled Seq\", \"S. cerevisiae \", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TATCATTTATCATTGAATGTATGTAATCATTTACCATGTGAGAATGTGTGATAGTAGGAGTAGCATATAATATTATCTATATATGTTATGTACAAAAATGGGCCTAGTATCTTAAGAATTCTGTTTATTTTAAATATCTCTGAACAAGTCAGATGAAGAGAGATTTGATCCGGGGCCGAGGACGGTCATGGGTGAATCTTTCTTTATTTTTTGTTTAGTCTGTTGTTTCCGAATTAGAGTCGTCTAGCATTGTTGTTCCGATGAGATGACGAATGTTATGGGGCAACATGTAGATTCAGCATATGTTTACTAAATAGGGGTGCGTTTTACTGTAACGTTAAGCGGTCGTTCGTGGGTTCCGGAAAGGCGGTTCACGTGGCAACGATCGGACGATACCAAGTCGCTGCGGTGTTCCTGCGGAATAAGCATAGGTAGATATAACACACCTCGTTTCTTTAGTGGCGCAAGATCGTACCTAAGACTGAATCTCCGCAAGTCGGTATTAGGTCGCCTACCATCGAAGCGCCGTTACGGACTAGTCTGTCGGCGTTTTTGGTTAATAGGCTTACTTGACAAGGAGAGCATGATTCAAGTTAATGATAACGCCATTGTAAGTAGTCATCCCATTTTGATTGGACAGAGTGCTGCCAGATTTGGGTCGAGTGCAAGGGATAATCACCCACTTGTTAGGTTGCGAATGGCTTAAGACGAAGCCATACTATCCTTCTCGGCTGTAGCTTCTTAGTTTTTCGTTACAGCGATACTTGCGAACTGACGGTGTTCGGTCAATAGGGACACCATTGAAAAGACCGTGGAGATCGGAGTTTAAGGCTCCCTGATTTCCTAGCTATCCGGTGTGAAAGTACCAAACATAAGTGTGACTTTTAGTTTTAGTTCCCCCGAAAATGGGAAAACAAGATGACCTCTAAAGACAAGAGGTACTCGGGGGGAATCCTGTAGACGCAATAGCAAATTGTCTACACGGCGGGGTCGGTTTGAGGGGTGGACTGTTACAGAAGTTGGGCCTAGTCGGGGCTTACCCTGGAACTTACGATCTTGCACCTTTTACTTAAGGTCGAGGCGAAGTAACTTATTCATTTCTTTGATATTTCCATCACCATAAAGTGACCGCGGCTTCGAGGGTGAATAAGATGTGGGAGATACAGAGAAGTGTTACAGTTTGATCTCAGTTCGAGTTGTCCCAGAAGAAAGGGGCGACTAAGACGGTTCGGGCAAGGGAACCGACACCAAAGCGATCTATCATCTATCCCTGTCACCCTTAGAGCAATTAGGTAAGTACGCGCAGTGATTAATCTACTGCTCCGTAAACCGATGGAATTCTCTCAGTATCAATGAGGGCGGCAAATGGGCGCGAACCAACTTAAAGAAGTGAAACTGTAAGTCTCGTGACCCGTCTTTAGTGTAACGCAGTTGTAGTGAAAGACTGGTAGCGTTACGATACAAAATTAATCTGTCAGTCTAAGGGGAACAGGCATGGTCAAGATTCAACTAGCAATTAACATCGTTCGCTGCCAGATGTTCTCTGGATGGTTCCGGCAGATGTTGTTCCGTGCGTTCATCAGGCGGATCGTCTCGTTCGGGGTGGTTCGTCAGGTGTTCGTGCGGGCGACGCAGACTGGTTCCGGGAGTGATGGGCTGGGAATCTCGGTTAGGAATAGGGCTTCAATGCCTAGATAAAACGGCTGAAGGGAATAGATGTAATAAGATAGTTGATCTCCGACAAGTGGAACCTCTGGACGACGCCAATAGTCATGCTGGACCGTACTTTTGATAAGGAAGGACACCTAAAAGTGCCCGGCAGTGTTCGCGTGGCCTCGGTCGTTTCCACGACCGGAGAAGGTCGGTATTCTGGGGTAGAGGCCTATTTGGTTAAGGCCCCACTATTCGACAATTCTTCTTTTCTATTGAGGAGGGTCCCGAGCGCGGCTGCAGAGGTGTAAGTCAATGCAATGGCACTTCTTAGGTATAGGTCCAAGGCCTTAGAATTGGCCTAAGGGAAAGCTACCACCGGACGTATTTTAGTCCGGAAACTTTGCCTCGAAGGGGTAGAGAATCCTAGCTGATTGGGTGCAGGTTGACGACAACTGCACCTTGGAAAGGGGTGAAGTCAGAAGTTTCAAGAGTAAACTTATAAACGATGATGGTGGTTCTAGACGTGATCTCCGGCAAGCTGGGCTGGAATGCCAGATCCGAAGCAGTGACTGGAGGTGCGGACGGATGAGCAGTCCCGTAGTATAGTTGGGACTGCCATCTCATATCCATTGTGCGAACTCGCGGTAGGTAAAAGTCCCGATCAAGTAAGCCGGCCACTCAACAATGTGTGAGGAATCGCCTAAGGCTGAAGGTACCGGTGGCAGGCCGACAGATCTACTTGATTGTGGAAAACACCACAGACTACTCGCACATAAGGCCGTGGAATTGAGATGCAAGCCAAGTAGGGCGTAGCGGTCAAGACGAATGGTTTTTACCGGGTGATTTTCGAGAAGTAAGTTTACAGGTGCAAGTTAATTCATTGTTCCTGAAGAATGTATAAATTTCAAACTCTTATCCAGTTCCAGTAAAGCTGGGGCCTTGGAGATTAGTAAGCGAAATGGAGTATTTTGACTATGCTCGAAGACGATAGGACTCCCTTTGAAGCCGTCCTTGGTCGATGATCTACCAAGCTAATCAGAAAGCGGGGATATGGGTTTAAGCTGCTAGCTAAACGTGCAGTCTTGGCGATGCTCGGAGGTGGTCTCAAAGGAGACCGAAGTGGGATAAGTCCGTATCAAGTGGTAGAAAGCCCAGGGTTGTCGATACGAGAATGAGTTTAGGTAGGCTTCTGTAGTCCTAGCCAGCTAACACGTGGAGAACGCTCCGGGGTTGGATGCAAGTGAAAGTAATGCGCATACCCAAAATGTGGGTTTGTGAGCGTATCTGCAATCTGAGGAACCAGGCACAAAGTTCTGCCCGCCGTATATTGGTAATACGGTCGTAGGAACTGAATGCAGCGTCAGGAGTCAGGGTCGACCGTCATAAGGGTGTCCGATATTATGAATGGCTCCGTTCGATGTAAGGATACCTAAATAGGACGGTGGTTTTGACTACGACCGGGTCACTTTACGCTCTAAGGGGATGGGTGTTCCTCGTCTCCCGTGTTTTGTGGTACAGACTAGTTTACGGGAAGGGAAAGTTGTTAAAGTGCATGAAAAAGTGAGAGAAAAGTTTCAAGAAAAGTAGAAAGGTAGTGACATGAACAAGCGATAGCCAGAGAGCGGTTATAAATCGAAATCTACCTTAAATGGTGGGTGAATCTCGACGTAAGGGTTTGTTGAGCTGAGAAGCTTCCGTGAAATGTTTCTTGGCGTGAGGAGCGGTGTGCCCTAAGAGTGGGAGATACTGCAGGACAAGGTTCCTTGTATCTGTTCCTTGCCGGGGTTTCAACGGGAGAGGTTTAATGTTGAGCCCGTGGCTTCCATGGTCTAAAGTTTAAACTCGAAAACGGCGAAGTGAGCGGCAATGATTCCGTTAGGGCCAACCAAAGAAAAGGAGGCGAATAACTATACGAATTCAAGTCGCCCATGAGGATGGACTAAACTCCAGTTTGAAATTCTTGTAACAAGCGGATCTGCGAGAGAAGAATAGCTATTGCAAGGTTATGCGAGTCATATTTTTTTCTAATCGGCGTCAACCATTTTGGATTTTGCTGGCATGAACGTAATATGGAGTTCGTGCGTCTCTTTGGAGAGAAACCTTTTTTTTGTAGGTTACTTTTCCGGTCGTTAAAGTTCAATTGAGGTTTCTCATAGTGAGTGATGGTTTGTCTTACAAACTCTTCCTTTACTGCGAGTTTGTCCGTACGGGGGACCTTATGGTTCCCCGCGTTACACGCAAGTTTCTAAGCTACTAAGTGCCTTAAGACGTTAAGTGTAATGCATAGCGTAAAGCGACGCAAGAAGTAGCTACGCTCTTGGTTCTCTAGGCAACAACTTTCAAAAATTATAAAATTTTAAAGGTCAATGCTTTTAAGAACAAAAACTGTTTTTAAATTACTTATCTATTTTAACAAACACAAACAATGGAGACCCGGGGCTAACGAGCTTACGGGTTTCTTTTTCAACGTTTCTATACTTTTGAGGNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTTTACTGTTTAGTGGAAAAAGTAGCATACGTGGAATAAGAGGTGTAGTGTTACGTGATAACGAAAACGAAAAAGTGGACAGTATAGGATAACGATAATCTACTTTATATTATTTTTAACAGGAGGTGGGTATTGTGGAGAGTGAGGGTGGATGACTTGTACAGACCTGGGACGGGAGTATAGTGGACGCAAAGGCAATTTGATAGCCAACGCCGGTATAGATGGTCTTTCGTGGCAAAGGGCAGGCTAGTTGACATCAATTCGACCATTCTCGGACTGGCTCATCACATCACCCACTGGTATGCGCTTTGAGTCCACGACGTCAACAAAAAAAAAAGCGCGTGCTTATGTAGTGCGTCGGAGCGGAACTGTGACGAGTTCCCTATTCCCCTTCCTCGACTCTATCTTGACTCAACCTCGCAACGTCGCCGTGCTCGTCACTTTTCTCGCTTCTCTTCTCTTCGTCTATTTTCTGGTTTCATCGCTACTATGAAGCTACAAACCTCGCTAAAAAGGATACACTTATAATCTTTAGTATACAAAACATAGTCGACTTGTCACTTACTACACGAGTTATGAGTTTAAGTAGAAAAACAAACGATTTTTATCGAATTGACCGTGTAGTACTTTTCCTACATCTTTAATATCTCATATAAATGATGATTCTTGATATATTTTTGTGTCACACTACTTTTCAGTGACTAGTCCGAAATTGGTGGCATCTGCGGTCTTTTTTGTCTGGATTAGACTAGGTACGGCATCGGCCTCATGTTTGTATTAACATCGTTTTATGCGCTAAAAACTCCCAAATCCATCGGTCTAGGTTACCGCATGAGACCCCGTAGTAGACCTCCTTGACATGGTAGACGACAAGAAGACGACACTTGTTAACCATATGGTATGAGAAGTATTTCAACAAGCAAATGTTATGATGGGCATCGTCAGGTCCTTGGTCGAGGTCTTGGGCGGTTGTGCTATGGAAAACGCGGGTCCAGGAACTGCCGGAAATGATAATCGAGTCCTAAGTTCGGAACTATCCTCATGTCTTATTACTGAAGACCTTAAAGGCTTCCCAGACACTCCAACCTTTACAGTTTTTTAACGGTAGGTAAACCGAACTGACCACCCATCATTTTGAGGTGCAGTAGTAAACTTTTCATTGGATTTATAGGGACAAGTAGGCGAGACTTACATAGATTTCTCAACCGTAAGTAAAAGCAACAGGTTTTGGGTCTCCGTTATGCTAGCAAATCTCACTAGTAGCACGGTGCTGGGTCACTAAAGAGTCTTCGTCGTTATCTGTGACGAACTATATTTAAGTAACCCGGTAGTCGTCTGCACCGTCACCGACCTGCGTATCGCGGACGCTATGTATGACCGAAGAGACTTAAATACCAGTTTAGATTCTTCTTTCGACAGAGGAGGTATCACAGCCAGGGTACTCACTGTTGGCGTGGTCGCATCAGTAGTCTCCGATCACGGAGCCTCTATGGCACTATGTTACCCTAGTCTACGCATATCAAATTTCTTAAACTTGGTTGCAATCTGTGAACTTCTATCAGGTCTAAACGGAAGAGTCGATTTCTACCCTGCCGAAGATAATTTAGTAAATGCCAATTAGGGTGCGATATTGGGCGGCAACAACGTGAACAGCTTGGGAACCTTCGCTATCATGGCGGCCATGGTTTTTAAAACTAACTACCGTTTCTTCTTAAGAAGAGAAGTGCTTAAAAATCATCACCTCGTGGTGACCTGTAACGCTGTTTCTCTCTTTCATTTTCCCATAAATTTCTAGAGTATCAACCTCCAACCACTCATGGTCTAACGAATGTTTTCTTATCGCTCGGTTGTAAACGAGACGGAGTCCGGAGAACCACGACGAACTTCTGAGTAGAATATACCGAAAACATACAGTACTAAACAAGAACATGTAATACACAACTATAATTTGTTTAACTAAAAAAAAAAAAACGCTATCGTTCGTCTATTACTTTCTCTGTTCCTGAACCTTGTAGGCTATTCTGACGCGGCTATAGCTAGAATGTCAGGAAGGGAACACAGTACTGAAAGCCTTTTCGTAGGAGCAGCTGACCATCAAACGACAGACAGTGCACGACTTCCCAGACTATGTAAAAAAATTTCTATTCTCTGCCCCAAATGGGAAGCCTCCTGATTCGCTCTAGAGGTTCATTTCTAGAGCGAATAGTTCTTTCGTCGGTTCACACCTTGCAGGAAAAAAACCAAAGTTTTTCTATAAGTAGTCAAATGTGACGTCGAAATTAACGGAGTTTTCCTATAGTACTCCACTAGATCCCAGTCTTCCCTTTCTAATGTCGTAGAACTCAACTTAGTGTAGACGTTTTCCACCATAATAACTGCAACGAGAAGGAATTACCTTTGAGTACCCCAAACCTTTCCTCCACGCCATTAGATAAAAAAAGCTTGTGTTTTGGATTGGAACTTTTCTTTGACAGGTTAAAGTAACTTGAATGGAGTCTTGCCCGGCCTCAGAAACGAAAGTCAGATTGTACCAGATTAAAGAAGCTTTTCGAAGTAAATTAACAATCTGACACCAAAATGTTCCTTTTTTGGTCACGATATGACTTCGCTATGGGTCTTGATTAATGGAACACACTGCTAAGCCGAGTCGCT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'   \\nprint(chrom12.find(\"ATCCTGCCAG\"))\\nprint(chrom12revcomp.find(\"ATCCTGCCAG\"))\\nrdn18start = chrom12revcomp.find(\"ATCCTGCCAG\") - 10\\nprint(chrom12revcomp[rdn18start:rdn18start+1801])\\n'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathname = Path(\"SC 7000071044\" + \"/\")\n",
    "fastafile = \"13003958421.gb\"\n",
    "inputfile = pathname / fastafile\n",
    "\n",
    "for seq_record in SeqIO.parse(inputfile, \"gb\"):\n",
    "    chrom12 = str(seq_record.seq[::])\n",
    "    chrom12rdn37revcomp = seq_record.seq[431334-1:438191]\n",
    "    rdn37 = chrom12rdn37revcomp.complement()\n",
    "    \n",
    "\n",
    "print(rdn37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_string = open(fastaoligname + \".fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", fasta_string)\n",
    "blast_result = open(\"Blast Total\" + fastaoligname + \".xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "Missing parentheses in call to 'print'. Did you mean print(\"***** RECORD ****\")? (<ipython-input-94-0e6c6039064e>, line 9)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-94-0e6c6039064e>\"\u001b[1;36m, line \u001b[1;32m9\u001b[0m\n\u001b[1;33m    print \"***** RECORD ****\"\u001b[0m\n\u001b[1;37m          ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m Missing parentheses in call to 'print'. Did you mean print(\"***** RECORD ****\")?\n"
     ]
    }
   ],
   "source": [
    "fasta_string = open(fastaoligname + \".fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", database=\"refseq_genomic\", sequence=fasta_string, entrez_query=\"txid4932[ORGN]\")\n",
    "records = NCBIXML.parse(result_handle)\n",
    "blast_record = records.next()\n",
    "\n",
    "for alignment in blast_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        if hsp.expect < 5:\n",
    "            print (\"***** RECORD ****\")\n",
    "            print \"sequence:\", alignment.title\n",
    "            print \"E-value:\", hsp.expect\n",
    "            \n",
    "blast_result = open(\"Blast \" + fastaoligname + \".xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "blastpath = Path(\"NCBI or Manual Compiling Testing/SC Testing/\")\n",
    "blastname = \"Blast_SC_CEN_PK113-7D.xml\"\n",
    "blastfile = blastpath / blastname\n",
    "result_handle = open(blastfile,\"r\")\n",
    "\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "blast_records = list(blast_records)\n",
    "E_VALUE_THRESH = 0.04\n",
    "for blast_record in blast_records:\n",
    "    print(blast_record.database_sequences)\n",
    "    for alignment in blast_record.alignments:\n",
    "        print(alignment)\n",
    "        for hsp in alignment.hsps:\n",
    "            print(hsp)\n",
    "            print(hsp.expect)\n",
    "            if hsp.expect < E_VALUE_THRESH:\n",
    "                print(\"****Alignment****\")\n",
    "                print(\"sequence:\", alignment.title)\n",
    "                print(\"length:\", alignment.length)\n",
    "                print(\"e value:\", hsp.expect)\n",
    "                print(hsp.query[0:75] + \"...\")\n",
    "                print(hsp.match[0:75] + \"...\")\n",
    "                print(hsp.sbjct[0:75] + \"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Blast in module Bio.Blast.Record object:\n",
      "\n",
      "class Blast(Header, DatabaseReport, Parameters)\n",
      " |  Saves the results from a blast search.\n",
      " |  \n",
      " |  Members:\n",
      " |  descriptions        A list of Description objects.\n",
      " |  alignments          A list of Alignment objects.\n",
      " |  multiple_alignment  A MultipleAlignment object.\n",
      " |  + members inherited from base classes\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Blast\n",
      " |      Header\n",
      " |      DatabaseReport\n",
      " |      Parameters\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self)\n",
      " |      Initialize the class.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Header:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(blast_record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "fasta_string = open(fastaoligname + \".fasta\").read()\n",
    "gbaccn = \"Z13005731920\"\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", \n",
    "                               database=\"refseq_genomic\", \n",
    "                               sequence=fasta_string, \n",
    "                               entrez_query=gbaccn+\"[ACCN]\")\n",
    "blast_result = open(gbaccn + \"Blast \" + fastaoligname + \".xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()\n",
    "\n",
    "fasta_string = open(\"Oligo Output.fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", database=\"refseq_genomic\", sequence=fasta_string, entrez_query=\"txid889517[ORGN]\")\n",
    "blast_result = open(\"Blast_SC_CEN_PK113-7D.xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
