{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import xlsxwriter\n",
    "import math\n",
    "import Bio\n",
    "import sys\n",
    "import os\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Alphabet import IUPAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skip the last oligo for RDN5, as it is shorter than or equal to 6 bp\n"
     ]
    }
   ],
   "source": [
    "\"\"\"DataFrame Initialization\n",
    "One for the original sequence reference\n",
    "and the other for oligos\n",
    "\n",
    "Oligo generation and original sequence reference recording\n",
    "Iterate through the number of sequences \n",
    "to update the original sequence reference dataframe with\n",
    "the sequence ID, the original sequence, \n",
    "its reverse and complementary sequence, and its sequence length \n",
    "\n",
    "For each sequence, iterate per 60 bp \n",
    "to update the oligo dataframe with\n",
    "the target name that is the sequence ID, \n",
    "the probe name that is the sequence ID \n",
    "plus the oligo number for that sequence ID,\n",
    "and the 60 bp or shorter oligo sequence itself \n",
    "in the 5' to 3' direction\n",
    "\n",
    "Note that oligos less than 4 bp are elimiated\n",
    "\n",
    "Export the dataframes to an Excel file\n",
    "One dataframe per sheet with adjustments such as\n",
    "the column widths, alignments, and text fonts.\n",
    "\"\"\"\n",
    "\n",
    "def SDRNA_Oligos(inputfolder, fastafile, strainname, details):\n",
    "    pathname = Path(inputfolder + \"/\")\n",
    "    fastafile = fastafile + \".fasta\"\n",
    "    inputfile = pathname / fastafile\n",
    "    \n",
    "    d = {\"SeqID\": \"\", \"ConSeq\": \"\", \"RevComp\": \"\", \"SeqLen\": []}\n",
    "    df = pd.DataFrame(data=d)\n",
    "\n",
    "    opsheet = {\"Target\": [],\n",
    "               \"Probe Name\": [],\n",
    "               \"Sequence (5'-3')\": []}\n",
    "    opf = pd.DataFrame(data=opsheet)\n",
    "    opname = (\"SDRNA oligos in \" + str(strainname) \n",
    "             + \" \" + str(details))\n",
    "    \n",
    "    opfasta = opname + \".fasta\"\n",
    "    opfastapath = pathname / opfasta\n",
    "    f = open(opfastapath, \"w\")\n",
    "\n",
    "    col1width = 0\n",
    "    col2width = 0 # initialize the column width for the probe name\n",
    "        \n",
    "    for seq_record in SeqIO.parse(inputfile, \"fasta\"):\n",
    "        revcomp = str(seq_record.seq.reverse_complement()[::])\n",
    "        seqlen = len(seq_record)\n",
    "        \n",
    "        targetname = seq_record.id.split(\"-\")[0]\n",
    "        # targetname = seq_record.id.split(\"|\")[4]\n",
    "                           \n",
    "        df = df.append({\"SeqID\": targetname,\n",
    "                        \"ConSeq\": str(seq_record.seq[::]),\n",
    "                        \"RevComp\": revcomp,\n",
    "                        \"SeqLen\": seqlen}, ignore_index=True)\n",
    "            \n",
    "        oligtot = math.ceil(seqlen/60)\n",
    "        \n",
    "        for olignumb in range(oligtot):\n",
    "                                  \n",
    "            seqstart = olignumb*60\n",
    "            \n",
    "            if olignumb != oligtot-1:\n",
    "                seqend = seqstart+60\n",
    "            elif olignumb == oligtot-1 and seqlen-seqstart < 7:\n",
    "                print(\"Skip the last oligo for \" + str(targetname) + \n",
    "                      \", as it is shorter than or equal to 6 bp\")\n",
    "                continue\n",
    "            else:\n",
    "                seqstart = seqlen-60\n",
    "                seqend = seqlen\n",
    "            \n",
    "            probename = str(targetname) + \"-SDRNA-\" + str(olignumb+1)\n",
    "            probeseq = revcomp[seqstart:seqend]\n",
    "            \n",
    "            opf = opf.append({\"Target\": targetname,\n",
    "                              \"Probe Name\": probename,\n",
    "                              \"Sequence (5'-3')\": probeseq}, ignore_index=True)\n",
    "\n",
    "            f.write(\">%s\\n%s\\n\" % (probename,probeseq))\n",
    "         \n",
    "        if len(targetname) > col1width:\n",
    "            col1width = len(targetname)\n",
    "        \n",
    "        if len(probename) > col2width:\n",
    "            col2width = len(probename)\n",
    "            \n",
    "    f.close()\n",
    "    \n",
    "    fastaoligname = \"SDRNA oligos in \" + str(strainname) + \" \" + str(details)\n",
    "    \n",
    "    df[\"SeqLen\"] = df[\"SeqLen\"].astype(\"int\")\n",
    "    \n",
    "    opExcel = opname + \".xlsx\"\n",
    "    opExcelpath = pathname / opExcel\n",
    "    writer = pd.ExcelWriter(opExcelpath, engine=\"xlsxwriter\")\n",
    "    opf.to_excel(writer, sheet_name=\"Oligos\", \n",
    "                 header=True, index=False, index_label=None)\n",
    "    df.to_excel(writer, sheet_name=\"Sequences\", \n",
    "                header=True, index=False, index_label=None)\n",
    "    workbook  = writer.book\n",
    "    leftformat = workbook.add_format({\"text_wrap\": True,\n",
    "                                      \"valign\": \"vcenter\",\n",
    "                                      \"font_name\": \"Courier New\"})\n",
    "    centerformat = workbook.add_format({\"text_wrap\": True,\n",
    "                                        \"align\": \"center\",\n",
    "                                        \"valign\": \"vcenter\",\n",
    "                                        \"font_name\": \"Courier New\"})\n",
    "    \n",
    "    worksheet1 = writer.sheets[\"Oligos\"]\n",
    "    worksheet1.set_column(\"A:A\", col1width*2, leftformat)\n",
    "    worksheet1.set_column(\"B:B\", col2width*2, leftformat)\n",
    "    worksheet1.set_column(\"C:C\", 80, leftformat)\n",
    "    \n",
    "    worksheet2 = writer.sheets[\"Sequences\"]\n",
    "    worksheet2.set_column(\"A:A\", col1width*2, centerformat)\n",
    "    worksheet2.set_column(\"D:D\", None, centerformat)\n",
    "    worksheet2.set_column(\"B:C\", 60, leftformat)\n",
    "\n",
    "    writer.save()\n",
    "    writer.close()\n",
    "    \n",
    "    return opfastapath, opname, col2width\n",
    "    \n",
    "opfastapath, opname, col2width = SDRNA_Oligos(\"Manual Compiling Testing/SC Testing\", \"SC Compiled Seq\", \"S. cerevisiae \", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "987591"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathname = Path(\"Manual Compiling Testing/SC Testing\" + \"/\")\n",
    "xmlfile = \"Full Blast \" + opname + \".xml\"\n",
    "xmlfilepath = pathname / xmlfile\n",
    "blastopname = \"Filtered Blast \" + opname\n",
    "blasttxt = blastopname + \".txt\"\n",
    "blasttxtpath = pathname / blasttxt\n",
    "blastExcel = blastopname + \".xlsx\"\n",
    "blastExcelpath = pathname / blastExcel\n",
    "blastop = {\"Oligo\": \"\", \"Target\": \"\", \"Align Length (bp)\": [], \"E-value\": [], \"Identities\": []}\n",
    "blastdf = pd.DataFrame(data=blastop)\n",
    "evalcut = 10**-5\n",
    "\n",
    "fasta_string = open(opfastapath).read()\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", database=\"refseq_genomic\", sequence=fasta_string, entrez_query=\"txid4932[ORGN]\")\n",
    "    \n",
    "blast_result = open(xmlfilepath, \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "t = open(blasttxtpath, \"w\")\n",
    "col2width = 0\n",
    "\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "for blast_record in blast_records:\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if ((hsp.expect < evalcut or hsp.align_length > 20) and \n",
    "                not (\"XII\" in alignment.title) and \n",
    "                not (\"rRNA\" in alignment.title)):\n",
    "                blastdf = blastdf.append({\"Oligo\": blast_record.query,\n",
    "                                          \"Target\": alignment.title,\n",
    "                                          \"Align Length (bp)\": hsp.align_length, \n",
    "                                          \"E-value\": hsp.expect,\n",
    "                                          \"Identities\": hsp.identities}, ignore_index=True)\n",
    "                if len(alignment.title) > col2width:\n",
    "                    col2width = len(alignment.title)\n",
    "                              \n",
    "                t.write((\"{0}\\n{1}\\n{2}\\n\\n\").format(blast_record.query, \n",
    "                                                     alignment.title, \n",
    "                                                     hsp))\n",
    "\n",
    "writer = pd.ExcelWriter(blastExcelpath, engine='xlsxwriter')                \n",
    "blastdf.to_excel(writer, header=True, index=False, index_label=None)\n",
    "\n",
    "worksheet = writer.sheets[\"Sheet1\"]\n",
    "\n",
    "worksheet.set_column(\"A:A\", col1width*2)\n",
    "    leftformat = workbook.add_format({\"text_wrap\": True,\n",
    "                                      \"valign\": \"vcenter\",\n",
    "                                      \"font_name\": \"Courier New\"})\n",
    "    \n",
    "\n",
    "\n",
    "t.close()\n",
    "writer.save()\n",
    "writer.close()\n",
    "# blast_result.close()\n",
    "# result_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "blastop = {\"Oligo\": \"\", \"Target\": \"\", \"Alignment Length\": [], \"E-value\": [], \"Identities\": []}\n",
    "blastdf = pd.DataFrame(data=blastop)\n",
    "evalcut = 10**-5\n",
    "\n",
    "blastopname = \"Filtered Blast \" + opname\n",
    "blasttxt = blastopname + \".txt\"\n",
    "blasttxtpath = pathname / blasttxt\n",
    "blastExcel = blastopname + \".xlsx\"\n",
    "blastExcelpath = pathname / blastExcel\n",
    "\n",
    "t = open(blasttxtpath, \"w\")\n",
    "blastcol2width = 0\n",
    "result_handle = open(xmlfilepath,\"r\")\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "for blast_record in blast_records:\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if ((hsp.expect < evalcut or hsp.align_length > 20) and \n",
    "                not (\"XII\" in alignment.title) and \n",
    "                not (\"rRNA\" in alignment.title)):\n",
    "                blastdf = blastdf.append({\"Oligo\": blast_record.query,\n",
    "                                          \"Target\": alignment.title,\n",
    "                                          \"Alignment Length\": hsp.align_length, \n",
    "                                          \"E-value\": hsp.expect,\n",
    "                                          \"Identities\": hsp.identities}, ignore_index=True)\n",
    "\n",
    "                if len(alignment.title) > blastcol2width:\n",
    "                    blastcol2width = len(alignment.title)\n",
    "                              \n",
    "                t.write((\"{0}\\n{1}\\n{2}\\n\\n\").format(blast_record.query, \n",
    "                                                     alignment.title, \n",
    "                                                     hsp))\n",
    "\n",
    "writer = pd.ExcelWriter(blastExcelpath, engine='xlsxwriter')  \n",
    "blastdf.to_excel(writer, header=True, index=False, index_label=None)\n",
    "\n",
    "workbook  = writer.book\n",
    "blastformat = workbook.add_format({\"font_name\": \"Courier New\", \n",
    "                                   \"align\": \"left\"})\n",
    "blastformat.set_font_size(10)\n",
    "worksheet = writer.sheets[\"Sheet1\"]\n",
    "worksheet.set_column(\"A:A\", col2width*1.2, blastformat)\n",
    "worksheet.set_column(\"B:B\", blastcol2width*1.1, blastformat)\n",
    "worksheet.set_column(\"C:C\", len(\"Alignment Length\"), blastformat)\n",
    "worksheet.set_column(\"D:E\", None, blastformat)\n",
    "\n",
    "t.close()\n",
    "writer.save()\n",
    "writer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "blastpath = Path(\"Manual Compiling Testing/SC Testing/\")\n",
    "blastname = \"Blast_SC_CEN_PK113-7D.xml\"\n",
    "blastfile = blastpath / blastname\n",
    "result_handle = open(blastfile,\"r\")\n",
    "\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "blast_records = list(blast_records)\n",
    "E_VALUE_THRESH = 0.04\n",
    "for blast_record in blast_records:\n",
    "    for alignment in blast_record.alignments:\n",
    "        print(alignment)\n",
    "        for hsp in alignment.hsps:\n",
    "            print(hsp)\n",
    "            print(hsp.expect)\n",
    "            if hsp.expect < E_VALUE_THRESH:\n",
    "                print(\"****Alignment****\")\n",
    "                print(\"sequence:\", alignment.title)\n",
    "                print(\"length:\", alignment.length)\n",
    "                print(\"e value:\", hsp.expect)\n",
    "                print(hsp.query[0:75] + \"...\")\n",
    "                print(hsp.match[0:75] + \"...\")\n",
    "                print(hsp.sbjct[0:75] + \"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_string = open(opfastapath).read()\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", fasta_string)\n",
    "blast_result = open(\"Blast Total\" + fastaoligname + \".xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "fasta_string = open(fastaoligname + \".fasta\").read()\n",
    "gbaccn = \"Z13005731920\"\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", \n",
    "                               database=\"refseq_genomic\", \n",
    "                               sequence=fasta_string, \n",
    "                               entrez_query=gbaccn+\"[ACCN]\")\n",
    "blast_result = open(gbaccn + \"Blast \" + fastaoligname + \".xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()\n",
    "\n",
    "fasta_string = open(\"Oligo Output.fasta\").read()\n",
    "result_handle = NCBIWWW.qblast(program=\"blastn\", database=\"refseq_genomic\", sequence=fasta_string, entrez_query=\"txid889517[ORGN]\")\n",
    "blast_result = open(\"Blast_SC_CEN_PK113-7D.xml\", \"w\")\n",
    "blast_result.write(result_handle.read())\n",
    "blast_result.close()\n",
    "result_handle.close()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
